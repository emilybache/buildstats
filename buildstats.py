#!/usr/bin/env python3

import datetime
import os
import re
import sys
from dataclasses import dataclass, field
from pathlib import Path


@dataclass(frozen=True, order=True)
class Build:
    """Contains the raw test_data from the log parsed into string fields,
    with methods to extract the processed test_data in more convenient formats"""
    when: str
    time_taken: str
    outcome: str
    tasks: str

    def to_csv(self):
        return f"{self.when}, {self.time_taken_in_seconds()}, {self.outcome}, {self.tasks}"

    def time_taken_in_seconds(self):
        return parse_to_secs(self.time_taken)


def parse_to_secs(raw_time):
    ms = 0
    secs = 0
    mins = 0
    hours = 0
    fields = [x.strip() for x in raw_time.strip().split()]
    if fields and fields[-1] == 'ms':
        ms = int(fields[-2])
        fields = fields[:-2]
    if fields and fields[-1] == 's':
        secs = int(fields[-2])
        fields = fields[:-2]
    if fields and fields[-1] == 'm':
        mins = int(fields[-2])
        fields = fields[:-2]
    if fields and fields[-1] == 'h':
        hours = int(fields[-2])
    t = datetime.timedelta(hours=hours, minutes=mins, seconds=secs, milliseconds=ms)
    return t.total_seconds()


@dataclass
class NamedRegex:
    regex: re.Pattern
    name: str


GRADLE_BUILD_RE = re.compile(r"^([\d\-:,\s]+) \[\d+\].* Gradle build (\w+) in ([\d\s\w]+)\n$")
GRADLE_TASKS_RE = re.compile(r"^.* About to execute Gradle tasks: \[([\w\s,:-]+)\].*$")


def next_match(lines, regexes):
    for line in lines:
        for named_regex in regexes:
            match = named_regex.regex.match(line)
            if match:
                yield named_regex.name, match


def next_build(matches):
    tasks = ""
    for name, match in matches:
        if name == "tasks":
            tasks = match.group(1)
        if name == "build":
            when = match.group(1)
            outcome = match.group(2)
            time_taken = match.group(3)
            yield Build(when=when, outcome=outcome, time_taken=time_taken, tasks=tasks)
            tasks = "" # reset tasks in case we get another build before another tasks


def filter_gradle_builds(lines):
    matches = next_match(lines, [NamedRegex(GRADLE_BUILD_RE, "build"), NamedRegex(GRADLE_TASKS_RE, "tasks")])
    builds = (build for build in next_build(matches))
    return builds


def output_filename(user=None, date=None):
    user = user or os.getlogin()
    date = date or datetime.date.today()
    return f"{date.isoformat()}-{user}.log"


def parse_builds(idea_log, output):
    for build in filter_gradle_builds(idea_log):
        output.write(f"{build}\n")


def parse_idea_log(log_file, output_filename):
    with open(log_file) as f:
        with open(output_filename, "a") as output_file:
            parse_builds(f, output_file)


def guess_path_to_idea_log():
    possible_paths = [
        Path.home() / "Library/Logs/Google/AndroidStudio4.2/idea.log",
        Path.home() / "Library/Logs/Google/AndroidStudio4.1/idea.log",
    ]
    for p in possible_paths:
        if p.exists():
            return p
    return None


def main(args):
    """
    Process the file created by the IDE into a log of builds,
    which it will put in the folder 'testdata'.
    By default it will look for the idea.log file in the default places for Android Studio versions 4.2 and 4.1.
    Pass an argument to look in a different place instead
    """
    if args:
        path = args[0]
    else:
        path = guess_path_to_idea_log()
    if not path:
        print(
            "unable to locate 'idea.log'! You can find it in your JetBrains IDE on the 'help' menu - 'Show log in Finder'. You should give the full path to idea.log as an argument to this script.")
        return

    data_folder = Path.cwd() / "testdata"
    if not data_folder.exists():
        os.mkdir(data_folder)
    output = data_folder / output_filename()
    print(f"Will parse log file {path} and write builds to {output}")
    parse_idea_log(path, output)


if __name__ == '__main__':
    main(sys.argv[1:])