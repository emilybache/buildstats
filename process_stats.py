#!/usr/bin/env python3
import math
import os
import statistics
import sys
import datetime
from pathlib import Path

from buildstats import Build, Sync


def total_time(builds):
    all_times = (b.time_taken_in_seconds() for b in builds)
    seconds = sum(all_times)
    return seconds


def total_build_count(builds):
    return len(builds)


def median_time(builds):
    all_times = (b.time_taken_in_seconds() for b in builds)
    median_time = statistics.median(all_times)
    return median_time


def deduplicate(builds):
    return list(set(builds))


def remove_clean_builds(builds):
    return (b for b in builds if b.is_a_clean_or_sync())


def gather_builds_from(f):
    all_builds = []
    for line in f:
        try:
            b = eval(line)
            all_builds.append(b)
        except NameError as e:
            print("Error gathering builds: ", e)
    return all_builds


def gather_builds(folder, output_csv):
    all_builds = []
    for root, directories, files in os.walk(folder):
        for file in files:
            if file.endswith(".log"):
                with open(os.path.join(root, file)) as log_file:
                    builds = gather_builds_from(log_file)
                    builds = deduplicate(builds)
                    all_builds.extend(builds)
    sorted_builds = sorted(all_builds, key=lambda x: x.when)
    for build in sorted_builds:
        output_csv.write(build.to_csv())
        output_csv.write("\n")
    return sorted_builds


def output_filename(user=None, date=None):
    user = user or os.getlogin()
    date = date or datetime.date.today()
    return f"{date.isoformat()}-{user}.csv"


def pretty_print_timedelta(seconds):
    days, seconds = divmod(seconds, 86400)
    hours, seconds = divmod(seconds, 3600)
    minutes, seconds = divmod(seconds, 60)
    milliseconds = round((seconds % 1) * 1000)
    if days > 0:
        return '%d d %d h %d m %d s %d ms' % (days, hours, minutes, seconds, milliseconds)
    elif hours > 0:
        return '%d h %d m %d s %d ms' % (hours, minutes, seconds, milliseconds)
    elif minutes > 0:
        return '%d m %d s %d ms' % (minutes, seconds, milliseconds)
    elif seconds >= 1:
        return '%d s %d ms' % (seconds, milliseconds)
    else:
        return '%d ms' % (milliseconds)


def summary_statistics(builds):
    builds_excluding_clean = [b for b in builds if not b.is_a_clean() and not b.is_a_sync()]
    syncs = [b for b in builds if b.is_a_sync()]
    summary = f"""
Builds (excluding 'clean')
----------------------------------
Total number: {total_build_count(builds_excluding_clean)}
Total time: {pretty_print_timedelta(total_time(builds_excluding_clean))}
Median time: {pretty_print_timedelta(median_time(builds_excluding_clean))}

Syncs
------
Total number: {total_build_count(syncs)}
Total time: {pretty_print_timedelta(total_time(syncs))}
Median time: {pretty_print_timedelta(median_time(syncs))}

All builds and syncs
--------------------
Total number: {total_build_count(builds)}
Total time: {pretty_print_timedelta(total_time(builds))}
Median time: {pretty_print_timedelta(median_time(builds))}
"""
    return summary


def main(args):
    """Process the log files created by the 'buildstats' script into a csv file,
    which it will put in the folder 'processed_data'.
    By default it will look for the log files in the folder 'data'.
    Pass an argument to look in a different folder instead"""
    if args:
        folder = args[0]
    else:
        folder = Path.cwd() / "data"
    processed_data_folder = Path.cwd() / "processed_data"
    if not processed_data_folder.exists():
        os.mkdir(processed_data_folder)
    output_path = processed_data_folder / output_filename()
    print(f"Will parse log files found under {folder} and write a csv file to {processed_data_folder}")
    with open(output_path, "w") as f:
        builds = gather_builds(folder, f)
        
    stats = summary_statistics(builds)
    sys.stdout.write(stats)


if __name__ == "__main__":
    main(sys.argv[1:])
