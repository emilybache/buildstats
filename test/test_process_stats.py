from io import StringIO

import pytest
from approvaltests import verify

from process_stats import *
from buildstats import Build
from process_stats import total_build_count_excluding_clean

SAMPLE_BUILDS = [
        Build(when='2021-07-16 13:58:29,316', time_taken='35 s 559 ms ', outcome='finished',
              tasks=':gmode:generateAcmeEuDebugSources, :gmode:createMockableJar, :gmode:generateAcmeEuDebugAndroidTestSources, :gmode:compileAcmeEuDebugUnitTestSources, :gmode:compileAcmeEuDebugAndroidTestSources, :gmode:compileAcmeEuDebugSources'),
        Build(when='2021-07-15 16:37:12,979', time_taken='3 m 28 s 451 ms ', outcome='finished', tasks='clean'),
        Build(when='2021-07-16 12:31:45,195', time_taken='2 m 22 s 428 ms ', outcome='finished', tasks=':app:assemble'),
        Build(when='2021-07-16 13:33:31,933', time_taken='2 m 17 s 252 ms ', outcome='finished', tasks=':app:assemble'),
    ]


def test_total_time():
    assert total_time(SAMPLE_BUILDS) == pytest.approx(523.69, 0.1)
    assert total_time_excluding_clean(SAMPLE_BUILDS) == pytest.approx(315.239, 0.1)


def test_total_build_count():
    assert total_build_count_excluding_clean(SAMPLE_BUILDS) == 3
    assert total_build_count(SAMPLE_BUILDS) == 4


def test_median_build_time():
    assert median_build_time(SAMPLE_BUILDS) == pytest.approx(139.84, 0.1)
    assert median_build_time_excluding_clean(SAMPLE_BUILDS) == pytest.approx(137.252, 0.1)


def test_deduplicate():
    builds = SAMPLE_BUILDS[:] + [Build(when='2021-07-16 12:31:45,195', time_taken='2 m 22 s 428 ms ', outcome='finished', tasks=':app:assemble')]
    assert len(deduplicate(builds)) == len(SAMPLE_BUILDS)


def test_gather_builds_from():
    file_contents = """\
Build(when='2021-07-15 16:54:09,703', time_taken='55 s 836 ms ', outcome='finished', tasks='clean')
Build(when='2021-07-15 16:59:35,922', time_taken='3 m 13 s 646 ms ', outcome='finished', tasks=':app:assembleAcmeEuDebug')
Build(when='2021-07-16 14:05:23,407', time_taken='31 s 669 ms ', outcome='finished', tasks=':gmode:generateAcmeEuDebugSources, :gmode:createMockableJar, :gmode:generateAcmeEuDebugAndroidTestSources, :gmode:compileAcmeEuDebugUnitTestSources, :gmode:compileAcmeEuDebugAndroidTestSources, :gmode:compileAcmeEuDebugSources')
"""
    builds = gather_builds_from(StringIO(file_contents))
    assert len(builds) == 3
    assert builds[0] == Build(when='2021-07-15 16:54:09,703', time_taken='55 s 836 ms ', outcome='finished', tasks='clean')


def test_gather_builds_from_with_junk():
    file_contents = """\
Build(when='2021-07-15 16:54:09,703', time_taken='55 s 836 ms ', outcome='finished', tasks='clean')
syntax_error
Build(when='2021-07-16 14:05:23,407', time_taken='31 s 669 ms ', outcome='finished', tasks=':gmode:generateAcmeEuDebugSources, :gmode:createMockableJar, :gmode:generateAcmeEuDebugAndroidTestSources, :gmode:compileAcmeEuDebugUnitTestSources, :gmode:compileAcmeEuDebugAndroidTestSources, :gmode:compileAcmeEuDebugSources')
"""
    builds = gather_builds_from(StringIO(file_contents))
    assert len(builds) == 2
    assert builds[0] == Build(when='2021-07-15 16:54:09,703', time_taken='55 s 836 ms ', outcome='finished', tasks='clean')


def test_gather_builds():
    folder = "data"
    output = StringIO()
    gather_builds(folder, output)
    verify(output.getvalue())