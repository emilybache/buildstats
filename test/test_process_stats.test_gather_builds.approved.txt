2021-07-15 12:31:45,195, 142.428, finished, :app:assembleAcmeEuDebug
2021-07-15 13:33:31,933, 137.252, finished, :app:assembleAcmeEuDebug
2021-07-16 12:31:45,195, 142.428, finished, :app:assemble
2021-07-16 13:33:31,933, 137.252, finished, :app:assemble
2021-07-16 13:58:29,316, 35.559, finished, :gmode:generateAcmeEuDebugSources, :gmode:createMockableJar, :gmode:generateAcmeEuDebugAndroidTestSources, :gmode:compileAcmeEuDebugUnitTestSources, :gmode:compileAcmeEuDebugAndroidTestSources, :gmode:compileAcmeEuDebugSources
2021-07-16 16:37:12,979, 208.451, finished, clean
